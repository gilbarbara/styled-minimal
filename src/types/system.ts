import * as React from 'react';
import * as StyledSystem from 'styled-system';
import * as StyledSystemCSS from '@styled-system/css';
import * as CSS from 'csstype';

import { Theme } from './theme';

export type ResponsiveSizes = '_' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | string;

export type ResponsiveObject = {
  [key: string]: CSSProps;
};

export interface OutlineProps {
  outline?: string;
  outlineColor?: string;
  outlineOffset?: number | string;
  outlineStyle?: string;
  outlineWidth?: number | string;
}

interface StyledSystemProps<T extends StyledSystem.Theme = StyledSystem.RequiredTheme>
  extends StyledSystem.BackgroundImageProps,
    // BackgroundProps
    StyledSystem.BackgroundPositionProps,
    StyledSystem.BackgroundRepeatProps,
    StyledSystem.BackgroundSizeProps,
    // BorderProps
    StyledSystem.BorderBottomProps,
    StyledSystem.BorderColorProps,
    StyledSystem.BorderLeftProps,
    StyledSystem.BorderRadiusProps,
    StyledSystem.BorderRightProps,
    StyledSystem.BorderStyleProps,
    StyledSystem.BorderTopProps,
    StyledSystem.BorderWidthProps,
    // ColorProps
    StyledSystem.BackgroundColorProps,
    StyledSystem.OpacityProps,
    // Flexbox
    StyledSystem.AlignContentProps,
    StyledSystem.AlignItemsProps,
    StyledSystem.AlignSelfProps,
    StyledSystem.FlexBasisProps,
    StyledSystem.FlexDirectionProps,
    StyledSystem.FlexGrowProps,
    StyledSystem.FlexProps,
    StyledSystem.FlexShrinkProps,
    StyledSystem.FlexWrapProps,
    StyledSystem.JustifyContentProps,
    StyledSystem.JustifyItemsProps,
    StyledSystem.JustifySelfProps,
    StyledSystem.OrderProps,
    // Grid
    StyledSystem.GridGapProps,
    StyledSystem.GridColumnGapProps,
    StyledSystem.GridRowGapProps,
    StyledSystem.GridColumnProps,
    StyledSystem.GridRowProps,
    StyledSystem.GridAutoFlowProps,
    StyledSystem.GridAutoColumnsProps,
    StyledSystem.GridAutoRowsProps,
    StyledSystem.GridTemplateColumnsProps,
    StyledSystem.GridTemplateRowsProps,
    StyledSystem.GridTemplateAreasProps,
    StyledSystem.GridAreaProps,
    // LayoutProps
    StyledSystem.DisplayProps,
    StyledSystem.HeightProps,
    StyledSystem.MaxHeightProps,
    StyledSystem.MaxWidthProps,
    StyledSystem.MinHeightProps,
    StyledSystem.MinWidthProps,
    StyledSystem.OverflowProps,
    StyledSystem.VerticalAlignProps,
    StyledSystem.WidthProps,
    // PositionProps
    StyledSystem.BottomProps,
    StyledSystem.LeftProps,
    StyledSystem.RightProps,
    StyledSystem.TopProps,
    StyledSystem.ZIndexProps,
    StyledSystem.SpaceProps,
    // TypographyProps
    StyledSystem.FontFamilyProps,
    StyledSystem.FontSizeProps,
    StyledSystem.FontStyleProps,
    StyledSystem.FontWeightProps,
    StyledSystem.LetterSpacingProps,
    StyledSystem.LineHeightProps,
    StyledSystem.TextAlignProps,
    StyledSystem.ShadowProps,
    OutlineProps {
  background?: StyledSystem.ResponsiveValue<CSS.BackgroundProperty<T>>;
  border?: StyledSystem.ResponsiveValue<CSS.BorderProperty<T>>;
  borderX?: StyledSystem.ResponsiveValue<CSS.BorderProperty<T>>;
  borderY?: StyledSystem.ResponsiveValue<CSS.BorderProperty<T>>;
  position?: StyledSystem.ResponsiveValue<CSS.PositionProperty>;
  textColor?: StyledSystem.ResponsiveValue<CSS.ColorProperty>;
}

export type CSSProps =
  | StyledSystemCSS.SystemStyleObject
  | Record<
      string,
      | StyledSystemCSS.SystemStyleObject
      | StyledSystemCSS.ResponsiveStyleValue<number | string>
      | Record<
          string,
          StyledSystemCSS.SystemStyleObject | StyledSystemCSS.ResponsiveStyleValue<number | string>
        >
    >;

export interface BaseProps extends React.RefAttributes<any>, StyledSystemProps {
  as?: React.ElementType;
  css?: CSSProps;
  cursor?: string;
  filter?: string;
  responsive?: ResponsiveObject;
  textDecoration?: string;
  textTransform?:
    | 'capitalize'
    | 'full-size-kana'
    | 'full-width'
    | 'inherit'
    | 'initial'
    | 'lowercase'
    | 'none'
    | 'unset'
    | 'uppercase';
  theme?: Theme;
}
