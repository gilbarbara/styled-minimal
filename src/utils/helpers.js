// @flow
import deepmerge from 'deepmerge';
import parseToRgb from 'polished/lib/color/parseToRgb';
import * as defaultTheme from './theme';

/** Check if a variable is defined **/
export const isDefined = (val: any): boolean %checks => typeof val !== 'undefined';

/** Check if a variable is a number **/
export const isNumber = (val: number | string): boolean %checks => typeof val === 'number';

/** Get textual unit value */
export const px = (val: string|number): string => (isNumber(val) ? `${val}px` : val);

const getValue = (value: any, key?: string): any => {
  if (key && ['size'].includes(key)) {
    return px(value);
  }

  return value;
};

/**
 *
 * @param {string|number} n
 * @param {Array} breakpoints
 * @returns {string}
 */
const createMediaQuery = (n: string | number, breakpoints: Array<number>): string => {
  const grid = {
    'xs-only': `@media (min-width: ${breakpoints[0] - 1}px)`,
    ix: `@media (min-width: ${breakpoints[0]}px)`,
    'ix-only': `@media (max-width: ${breakpoints[1] - 1}px)`,
    md: `@media (min-width: ${breakpoints[1]}px)`,
    'md-only': `@media (max-width: ${breakpoints[2] - 1}px)`,
    lg: `@media (min-width: ${breakpoints[2]}px)`,
    'lg-only': `@media (max-width: ${breakpoints[3] - 1}px)`,
    xl: `@media (min-width: ${breakpoints[3]}px)`,
    'xl-only': `@media (max-width: ${breakpoints[4] - 1}px)`,
    xxl: `@media (min-width: ${breakpoints[4]}px)`,
  };

  if (isNumber(n)) {
    return `@media (min-width: ${px(n)})`;
  }

  return grid[n] || `@media (min-width: ${n})`;
};

/**
 * Get Styles
 * @param {Object} props
 *
 * @returns {Object}
 */
export const getTheme = (props: Object = {}): Object => {
  const { theme = {} } = props;

  return deepmerge(defaultTheme, theme, { arrayMerge: (dest, source) => source });
};

/**
 * Returns the value of `props[path]` or `defaultValue`
 * @example
 * const Button = styled.button`
 *  color: ${prop('color', 'red')};
 * `
 */
/* eslint-disable react/destructuring-assignment */
export const themeGet = (props: Object, path?: string, options: Object = {}): any => {
  const theme = getTheme(props);
  const { base, key, toggle } = options;
  const selection = theme[path];
  let isActive = true;

  if (!path) {
    return theme;
  }

  if (typeof toggle !== 'undefined') {
    isActive = props[toggle] || false;
  }

  if (!isActive) {
    return base;
  }

  if (key) {
    if (Array.isArray(key)) {
      const value = key
        .filter(d => !!props[d])
        .map(d => props[d])[0] || base;

      return selection[value];
    }

    return getValue(selection[props[key] || base], key);
  }

  return getValue(selection);
};

/**
 * Get the space from its scale
 *
 * @param {number|string|Array} value
 * @param {boolean} [pure]
 *
 * @returns {function}
 */
export const spacer = (value: number | string | Array<number>, pure?: Boolean): any => (props: Object): any => {
  const { space } = themeGet(props);
  const result = space[value] || value;

  return pure ? result : px(result);
};

/**
 * Generate responsive media queries
 *
 * @param {function|Object} input
 * @param {function} [queryBuilderFn]
 *
 * @returns {function}
 */
export const responsive = (input: Function | Object, queryBuilderFn?: Function): Function => (props: Object): string => {
  const { breakpoints } = themeGet(props);
  const rules = typeof input === 'function' ? input(props) : input;
  const queryBuilder = queryBuilderFn || createMediaQuery;
  const result = [];

  for (const rule in rules) {
    /* istanbul ignore else */
    if ({}.hasOwnProperty.call(rules, rule)) {
      result.push(`
        ${queryBuilder(rule, breakpoints)} {
          ${rules[rule]}
        }
      `);
    }
  }

  return result.join('\n');
};

/**
 * Get Color from theme
 *
 * @param {Object} props
 * @param {string} base
 * @returns {*}
 */
export function getColor(props: Object, base: string = 'primary'): string {
  const { variant } = props;
  const colors = themeGet(props, 'colors');
  const palette = themeGet(props, 'palette');

  return palette[variant] || colors[variant] || palette[base];
}

/**
 * Color contrast
 *
 * @param {string} color
 * @returns {number}
 */
export function getYiq(color: string): number {
  const r = parseToRgb(color).red;
  const g = parseToRgb(color).green;
  const b = parseToRgb(color).blue;

  return (r * 299 + g * 587 + b * 114) / 1000;
}
