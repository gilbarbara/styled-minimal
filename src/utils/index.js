// @flow
import { colors, palette } from '../styles';

export const defaults = {
  alignHorizontal: ['left', 'center', 'right'],
  buttonTypes: ['button', 'submit', 'reset'],
  headingSizes: ['', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'],
  headingTypes: ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'],
  sizes: ['xs', 'sm', 'md', 'lg', 'xl'],
  variants: [...Object.keys(palette), ...Object.keys(colors)],
  weights: [
    'light',
    'normal',
    'medium',
    'bold',
  ],
};

export * from './helpers';
export * from './system';
