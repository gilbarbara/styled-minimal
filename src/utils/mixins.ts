export function placeholder(content: string): string {
  return `
    &::-webkit-input-placeholder {
      ${content};
    }
  
    &:-moz-placeholder {
      ${content};
    }
  
    &::-moz-placeholder {
      ${content};
    }
  
    &:-ms-input-placeholder {
      ${content};
    }
  `;
}
