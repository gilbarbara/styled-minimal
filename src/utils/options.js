// @flow
import { colors, palette } from '../styles';

export const propsOptions = {
  sizes: ['xs', 'sm', 'md', 'lg', 'xl'],
  variants: [...Object.keys(palette), ...Object.keys(colors)],
};

export const inputTextTypes = [
  'date',
  'email',
  'number',
  'password',
  'search',
  'tel',
  'text',
];
