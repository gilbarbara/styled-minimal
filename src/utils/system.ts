import {
  css,
  CSSObject,
  FlattenInterpolation,
  FlattenSimpleInterpolation,
  ThemedStyledProps,
} from 'styled-components';
import systemCSS from '@styled-system/css';

import {
  getColor,
  getDarkColor,
  getProps,
  getTextColor,
  getTheme,
  isDefined,
  px,
  responsive,
} from './helpers';
import { placeholder } from './mixins';
import { darken, desaturate, lighten } from './polished';
import { styleSystem } from './styles';

import { CommonProps, InputTypes } from '../types';

export const inputTextTypes: InputTypes[] = [
  'date',
  'email',
  'file',
  'number',
  'password',
  'search',
  'tel',
  'text',
];

export const baseStyles: CSSObject = {
  boxSizing: 'border-box',
};

export const colorStyles = (props: CommonProps): FlattenSimpleInterpolation => {
  const { dark, invert } = props;

  return css`
    color: ${invert || dark ? getColor(props) : getTextColor(props)};
  `;
};

export const lineHeightStyles = (props: CommonProps): number => getTheme(props, 'lineHeight');

export const variantStyles = (props: CommonProps): FlattenSimpleInterpolation => {
  const { dark, invert } = props;
  const { darkColor } = getTheme(props);
  const themeColor = getColor(props);

  const backgroundColor = invert ? 'transparent' : themeColor;
  const color = invert ? themeColor : getTextColor(props);

  return css`
    background-color: ${dark ? darkColor : backgroundColor};
    border: ${`1px solid ${dark ? darkColor : themeColor}`};
    color: ${dark ? getDarkColor(props) : color};
  `;
};

export const pseudoStyles = (
  props: CommonProps,
): FlattenInterpolation<ThemedStyledProps<any, any>> => {
  const {
    invert,
    outline,
    outlineColor,
    outlineOffset = 1,
    outlineStyle,
    outlineWidth,
    multiple,
    valid,
  } = props;
  const {
    backgroundColor,
    borderColor,
    color,
    disabledColor,
    requiredColor,
    validation,
  } = getTheme(props, 'input');

  const currentColor = color;
  let currentBorderColor = borderColor;

  if (valid) {
    currentBorderColor = validation.valid;
  } else if (valid === false) {
    currentBorderColor = validation.invalid;
  }

  const inputOnly = (): FlattenSimpleInterpolation | undefined => {
    let styles;

    /* istanbul ignore else */
    if (!isDefined(multiple)) {
      try {
        styles = css`
          ${placeholder(`color: ${lighten(0.5, currentColor)};`)};

          &:read-only {
            background-color: ${darken(0.02, backgroundColor)};
            color: ${lighten(0.3, currentColor)};
          }
        `;
      } catch (error) {
        // what to do?
      }
    }

    return styles;
  };

  const disabled = (): FlattenSimpleInterpolation | undefined => {
    let styles;

    /* istanbul ignore else */
    if (!isDefined(multiple)) {
      try {
        styles = css`
          ${placeholder(`color: ${lighten(0.5, currentColor)};`)};

          &:disabled {
            background-color: ${disabledColor};
            color: ${lighten(0.3, currentColor)};
            border-color: ${disabledColor};
          }
        `;
      } catch (error) {
        // what to do?
      }
    }

    return styles;
  };

  return css`
    ${inputOnly};
    ${disabled};
    ${placeholder(`color: ${lighten(0.5, currentColor)};`)};

    &:focus {
      ${outline ? `outline: ${outline}` : null};
      outline-color: ${outlineColor || (invert ? currentBorderColor : backgroundColor)};
      ${outlineOffset ? `outline-offset: ${px(outlineOffset)}` : null};
      ${outlineStyle ? `outline-style: ${outlineStyle}` : null};
      ${outlineWidth ? `outline-width: ${px(outlineWidth)}` : null};
    }

    &:required:not(:valid) {
      border-color: ${requiredColor};
      ${placeholder(`color: ${lighten(0.1, desaturate(0.2, requiredColor))};`)};
    }
  `;
};

export const sx = (props: any) => {
  const mergedProps = getProps(props);

  return {
    ...systemCSS(props.css)(mergedProps),
    ...responsive(props.responsive)(mergedProps),
  };
};
export const system = (props: any) => styleSystem(getProps(props));
