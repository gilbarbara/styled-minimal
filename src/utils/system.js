import lighten from 'polished/lib/color/lighten';
import darken from 'polished/lib/color/darken';
import { css } from 'styled-components';
import {
  border,
  fontFamily,
  fontSize,
  fontWeight,
  space,
  textAlign,
} from 'styled-system';

import { getColor, getProp, getYiq } from './helpers';
import { placeholder } from './mixins';
import { inputTextTypes } from './options';

export const systemGroups = (...opts) => {
  const groups = {
    flex: [
      'alignContent',
      'alignItems',
      'display',
      'flex',
      'flexDirection',
      'flexWrap',
      'justifyContent',
    ],
    layout: [
      'alignSelf',
      'display',
      'flex',
      'flexBasis',
      'height',
      'minHeight',
      'minWidth',
      'order',
      'space',
      'width',
    ],
    positioning: [
      'bottom',
      'left',
      'position',
      'right',
      'top',
      'zIndex',
    ],
    text: [
      'borders',
      'display',
      'fontFamily',
      'fontSize',
      'fontWeight',
      'space',
    ],
    ui: [
      'backgroundImage',
      'bgColor',
      'borders',
      'color',
      'fontFamily',
      'fontSize',
      'fontWeight',
      'textAlign',
    ],
  };

  return opts
    .filter(d => !!groups[d])
    .map(d => groups[d])
    .reduce((acc, data) => {
      data.forEach(d => {
        if (!acc.includes(d)) {
          acc.push(d);
        }
      });

      return acc;
    }, [])
    .sort((a, b) => {
      const left = a.toUpperCase();
      const right = b.toUpperCase();

      if (left < right) {
        return -1;
      }
      if (left > right) {
        return 1;
      }

      return 0;
    });
};

export const base = {
  color: props => {
    const { disabled, outline } = props;
    const colors = getProp('colors')(props);
    const selectedColor = getColor(props);

    let baseColor = getYiq(selectedColor) > 180 ? colors.black : colors.white;
    baseColor = outline ? selectedColor : baseColor;

    return css`color: ${disabled ? lighten(0.3, baseColor) : baseColor};`;
  },
  variant(props) {
    const { disabled, outline } = props;
    const colors = getProp('colors')(props);
    const selectedColor = getColor(props);
    const backgroundColor = outline ? colors.white : getColor(props);
    const borderColor = disabled ? lighten(0.3, selectedColor) : selectedColor;
    let color = getYiq(selectedColor) > 180 ? colors.black : colors.white;
    color = outline ? selectedColor : color;

    return css`
      background-color: ${disabled && !outline ? lighten(0.2, backgroundColor) : backgroundColor};
      border: ${outline ? `1px solid ${borderColor}` : 0};
      color: ${disabled ? lighten(0.3, color) : color};
    `;
  },
  fontFamily: props => getProp('fontFamily')(props),
  fontSize: props => getProp('fontSizes', { key: 'size', base: 'md' })(props),
  fontWeight: props => getProp('fontWeights', { key: 'weight', base: 'normal' })(props),
  lineHeight: props => getProp('lineHeight')(props),
};

export const AlertStyles = {
  base(props) {
    const { size } = props;
    const { borderRadius, maxWidth, padding } = getProp('alert')(props);

    return css`
      ${base.variant};
      border-radius: ${borderRadius};
      font-family: ${base.fontFamily};
      font-size: ${base.fontSize};
      line-height: ${base.lineHeight};
      max-width: ${maxWidth};
      padding: ${padding[size]};
      width: 100%;
      
      a {
        ${base.color};
      }
    `;
  },
};

export const BadgeStyles = {
  base(props) {
    const {
      borderRadius,
      fontSize: badgeFontSize,
      fontWeight: badgeFontWeight,
      padding,
    } = getProp('badge')(props);
    const fontSizeProp = getProp('fontSizes', { key: 'size' })(props);

    return css`
      align-items: center;
      ${base.variant};
      border-radius: ${borderRadius};
      display: inline-flex;
      font-size: ${badgeFontSize};
      ${fontSizeProp ? `font-size: ${fontSizeProp}` : ''};
      font-weight: ${badgeFontWeight};
      line-height: ${base.lineHeight};
      padding: ${padding};
      vertical-align: baseline;
    `;
  },
};

export const ButtonStyles = {
  base(props) {
    const { animate, outline, size } = props;
    const {
      borderRadius,
      lineHeight: buttonLineHeight,
      loader,
      padding,
    } = getProp('button')(props);
    const fontSizeProp = getProp('fontSizes', { key: 'size' })(props);

    return css`
      align-items: center;
      ${base.variant};
      border-radius: ${borderRadius[size]};
      box-shadow: none;
      cursor: pointer;
      display: inline-flex;
      font-family: ${base.fontFamily};
      ${fontSizeProp ? `font-size: ${fontSizeProp}` : ''};
      font-weight: ${base.fontWeight};
      justify-content: center;
      line-height: ${buttonLineHeight};
      padding: ${padding[size]};
      text-decoration: none;
      width: ${({ block }) => (block ? '100%' : 'auto')}
      ${animate ? loader(outline ? '#ccc' : '#fff') : ''};
    `;
  },
  outlineColor(props) {
    return css`outline-color: ${getColor(props)};`;
  },
};

export const ContainerStyles = {
  base(props) {
    const containerMaxWidth = getProp('containerMaxWidth')(props);

    return css`
      margin-left: auto;
      margin-right: auto;
      max-width: ${containerMaxWidth};
      position: relative;
      width: 100%;
    `;
  },
};

export const FormStyles = {
  fieldset(props) {
    const {
      backgroundColor: bgColor,
      borderColor,
      borderRadius,
      padding,
      marginBottom,
    } = getProp('fieldset')(props);

    return css`
      background-color: ${bgColor};
      border: 1px solid ${borderColor};
      border-radius: ${borderRadius};
      margin-bottom: ${marginBottom};
      padding: ${padding};
      text-align:left;
    `;
  },
  form(props) {
    const { bordered } = props;
    const {
      backgroundColor: formBgColor,
      borderColor,
      borderRadius,
      padding,
    } = getProp('form')(props);

    return css`
      background-color: ${formBgColor};
      ${bordered ? `border: 1px solid ${borderColor};` : ''}
      ${bordered ? `border-radius: ${borderRadius};` : ''}
      ${bordered ? `padding: ${padding};` : ''}
      text-align:left;
    `;
  },
  group(props) {
    const { bordered } = props;
    const {
      backgroundColor: formGroupBgColor,
      borderColor,
      borderRadius,
      marginBottom,
      padding,
    } = getProp('formGroup')(props);

    return css`
      background-color: ${formGroupBgColor};
      ${bordered ? `border: 1px solid ${borderColor};` : ''}
      ${bordered ? `border-radius: ${borderRadius};` : ''}
      margin-bottom: ${marginBottom};
      ${bordered ? `padding: ${padding};` : ''};
      text-align: left;
    `;
  },
  input(props) {
    const { size, type, valid } = props;
    const {
      backgroundColor: inputBgColor,
      borderColor,
      borderRadius,
      borderWidth,
      color: inputColor,
      fontSize: inputFontSize,
      height,
      inlineMargin,
      lineHeight: inputLineHeight,
      padding,
      validation,
    } = getProp('input')(props);
    let thisColor = borderColor;

    if (valid) {
      thisColor = validation.valid;
    }
    else if (valid === false) {
      thisColor = validation.invalid;
    }

    return css`
      background-color: ${inputBgColor};
      border: ${borderWidth ? `${borderWidth} solid ${thisColor}` : ''};
      border-radius: ${borderRadius};
      color: ${inputColor};
      display: ${!['checkbox', 'radio'].includes(type) ? 'block' : 'inline-block'};
      font-family: ${base.fontFamily};
      font-size: ${inputFontSize[size]};
      ${inputTextTypes.includes(type) ? `height: ${height[size]}` : ''};
      line-height: ${inputLineHeight};
      ${['checkbox', 'radio'].includes(type) ? `margin: 0 ${inlineMargin} 0 0` : ''};
      padding: ${inputTextTypes.includes(type) ? padding[size] : 0};
      ${!['checkbox', 'radio', 'color'].includes(type) ? 'width: 100%;' : ''};
    `;
  },
  label(props) {
    const { inline } = props;
    const { color: labelColor, inlineFontSize, marginBottom } = getProp('label')(props);

    return css`
      color: ${labelColor};
      display: block;
      font-family: ${base.fontFamily};
      ${inline ? `font-size: ${inlineFontSize}` : ''};
      line-height: ${base.lineHeight};
      ${!inline ? `margin-bottom: ${marginBottom}` : ''};
      white-space: nowrap;
    `;
  },
  legend(props) {
    const { color: legendColor, marginBottom } = getProp('legend')(props);

    return css`
      color: ${legendColor};
      display: block;
      font-family: ${base.fontFamily};
      line-height: ${base.lineHeight};
      margin-bottom: ${marginBottom};
      white-space: nowrap;
    `;
  },
  select(props) {
    const { multiple, sizing } = props;
    const {
      backgroundColor: bgColor,
      color: selectColor,
      borderColor,
      borderRadius,
      fontSize: selectFontSize,
      height,
      lineHeight: selectLineHeight,
      padding,
    } = getProp('select')(props);

    return css`
      background-color: ${bgColor};
      border: 1px solid ${borderColor};
      border-radius: ${borderRadius};
      color: ${selectColor};
      display: block;
      ${!multiple ? `height: ${height[sizing]}` : ''};
      font-family: ${base.fontFamily};
      font-size: ${selectFontSize[sizing]};
      line-height: ${selectLineHeight};
      ${!multiple ? `padding: ${padding[sizing]};` : ''};
      white-space: nowrap;
      width: 100%;
      
      > option {
        background-color: ${bgColor};
        font-size: ${selectFontSize[sizing]};
        line-height: ${selectLineHeight};
        padding: ${padding[sizing]};
      }
      
      &[multiple]:focus option:checked {
        background: black linear-gradient(0deg, black 0%, black 100%);
      }
    `;
  },
  textarea(props) {
    const { size, type, valid } = props;
    const {
      backgroundColor: inputBgColor,
      borderColor,
      borderRadius,
      borderWidth,
      color: inputColor,
      fontSize: inputFontSize,
      lineHeightTextarea,
      padding,
      validation,
    } = getProp('input')(props);
    let thisColor = borderColor;

    if (valid) {
      thisColor = validation.valid;
    }
    else if (valid === false) {
      thisColor = validation.invalid;
    }

    return css`
      background-color: ${inputBgColor};
      border: ${borderWidth ? `${borderWidth} solid ${thisColor}` : ''};
      border-radius: ${borderRadius};
      color: ${inputColor};
      display: ${!['checkbox', 'radio'].includes(type) ? 'block' : 'inline-block'};
      font-family: ${base.fontFamily};
      font-size: ${inputFontSize[size]};
      line-height: ${lineHeightTextarea};
      margin: 0;
      padding: ${padding[size]};
      width: 100%;
    `;
  },
  // helpers
  helpBlock(props) {
    const { helpColor, helpMarginTop } = getProp('formGroup')(props);

    return css`
      color: ${helpColor};
      display: block;
      font-family: ${base.fontFamily};
      font-size: 85%;
      line-height: 1.3;
      margin-top: ${helpMarginTop};
    `;
  },
  inlineMargin(props) {
    const { inlineMargin } = getProp('formGroup')(props);

    return css`margin-right: ${inlineMargin};`;
  },
  pseudo(props) {
    const { multiple } = props;
    const {
      backgroundColor: inputBgColor,
      color: inputColor,
      focusColor,
      requiredColor,
    } = getProp('input')(props);
    const inputOnly = typeof multiple === 'undefined'
      ? `
    ${placeholder(`color: ${lighten(0.5, inputColor)};`)};
    
    &:read-only {
      background-color: ${darken(0.02, inputBgColor)};
      color: ${lighten(0.3, inputColor)};
    }`
      : '';

    return css`
      ${inputOnly};
      
      &:disabled {
        background-color: ${darken(0.05, inputBgColor)};
        color: ${lighten(0.2, inputColor)};
      }
      
      &:focus {
        outline-color: ${focusColor};
      }
      
      &:required:not(:valid) {
        border-color: ${requiredColor};
        ${placeholder(`color: ${requiredColor};`)};
      }
    `;
  },
};

export const HeadingStyles = {
  base(props) {
    const { gutterBottom } = props;
    const headingWeight = getProp('headingWeight')(props);
    const headingGutter = getProp('headingGutter')(props);
    const headingSize = getProp('headingSizes', { key: ['size', 'is'], base: 'h1' })(props);

    return css`
      font-size: ${headingSize};
      ${border};
      font-family: ${base.fontFamily};
      font-weight: ${headingWeight};
      ${fontWeight};
      line-height: ${base.lineHeight};
      margin: 0 0 ${gutterBottom ? headingGutter : 0};
      ${space};
      ${textAlign};
    `;
  },
};

export const ListStyles = {
  base(props) {
    const { bordered, inline, is, styleType } = props;
    const { borderColor, borderRadius } = getProp('list')(props);

    return css`
      ${bordered ? `border: 1px solid ${borderColor};` : ''};
      ${bordered ? `border-radius: ${borderRadius};` : ''};
      display: flex;
      flex-direction: ${inline ? 'row' : 'column'};
      ${fontFamily};
      ${fontSize};
      ${fontWeight};
      margin: 0;
      ${is === 'ul' && styleType === 'none' ? 'padding: 0;' : ''};
      ${space};
      ${textAlign};
      ${is === 'ul' ? `list-style-type: ${styleType};` : ''};
    `;
  },
  itemBorder(props) {
    const { bordered } = props;
    const { borderColor } = getProp('list')(props);

    if (bordered) {
      return css`border-top: 1px solid ${borderColor};`;
    }

    return '';
  },
  itemPadding(props) {
    const { size } = props;
    const { padding } = getProp('list')(props);

    return css`padding: ${padding[size]};`;
  },
};

export const TableStyles = {
  base(props) {
    const { bordered, borderless, inverted } = props;
    const { borderColors, colors } = getProp('table')(props);

    return css`
      background-color: ${colors[inverted ? 'secondary' : 'primary']};
      ${bordered && !borderless ? `border: 1px solid ${borderColors[inverted ? 'secondary' : 'primary']};` : ''};
      border-collapse: collapse;
      color: ${colors[inverted ? 'primary' : 'secondary']};
      font-family: ${base.fontFamily};
      width: 100%;
    `;
  },
  caption(props) {
    const { captionColor, captionPadding } = getProp('table')(props);

    return css`
      caption {
        caption-side: bottom;
        color: ${captionColor};
        font-size: 90%;
        padding-bottom: ${captionPadding};
        padding-top: ${captionPadding};
        text-align: left;
      }
    `;
  },
  cellBorder(props) {
    const { bordered, borderless, inverted } = props;
    const { borderColors } = getProp('table')(props);

    if (borderless) {
      return '';
    }

    return css`${bordered ? 'border' : 'border-top'}: 1px solid ${borderColors[inverted ? 'secondary' : 'primary']};`;
  },
  color(props) {
    const { inverted } = props;
    const { colors } = getProp('table')(props);

    return css`color: ${colors[inverted ? 'primary' : 'secondary']};`;
  },
  headBackgroundColor(props) {
    const { head } = props;
    const { headColors } = getProp('table')(props);

    return css`background-color: ${headColors[head] || 'transparent'};`;
  },
  headCellBorder(props) {
    const { bordered, borderless, inverted } = props;
    const { borderColors } = getProp('table')(props);
    const colorProp = borderColors[inverted ? 'secondary' : 'primary'];

    if (borderless) {
      return '';
    }

    return css`
      ${bordered ? `border: 1px solid ${colorProp}` : `border-bottom: 2px solid ${colorProp}`};
      ${bordered && 'border-bottom-width: 2px'};
    `;
  },
  headColor(props) {
    const { head } = props;
    const { colors } = getProp('table')(props);

    if (head) {
      return css`color: ${colors[head === 'dark' ? 'primary' : 'secondary']};`;
    }

    return '';
  },
  padding(props) {
    const { size } = props;
    const { padding } = getProp('table')(props);

    return css`padding: ${padding[size]};`;
  },
  striped(props) {
    const { inverted, striped } = props;
    const { stripedColors } = getProp('table')(props);

    if (striped) {
      return css`background-color: ${stripedColors[inverted ? 'secondary' : 'primary']};`;
    }

    return '';
  },
};
