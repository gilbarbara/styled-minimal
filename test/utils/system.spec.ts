import { colorStyles, variantStyles, sx, system } from '../../src/utils/system';

describe('colorStyles', () => {
  it('should return the proper color', () => {
    expect(colorStyles({ variant: 'secondary' })).toMatchSnapshot();
    expect(colorStyles({ variant: 'yellow' })).toMatchSnapshot();
    expect(colorStyles({ variant: 'not-valid' })).toMatchSnapshot();
  });
});

describe('variantStyles', () => {
  it('should return the proper styles', () => {
    expect(variantStyles({ variant: 'secondary' })).toMatchSnapshot();
    expect(variantStyles({ variant: 'yellow' })).toMatchSnapshot();
    expect(variantStyles({ variant: 'not-valid' })).toMatchSnapshot();
  });
});

describe('sx', () => {
  it('should handle the css prop', () => {
    expect(
      sx({
        css: {
          bg: '#000',
          color: ['#f04', '#999'],
          '&:hover': { color: '#333' },
        },
      }),
    ).toMatchSnapshot();
  });

  it('should handle the responsive prop', () => {
    expect(
      sx({
        responsive: {
          _: {
            color: '#ccc',
          },
          md: {
            color: '#999',
          },
        },
      }),
    ).toMatchSnapshot();
  });
});

describe('system', () => {
  it('should handle props', () => {
    expect(
      system({ color: { _: '#333', sm: '#000' }, fontSize: [20, 22], fontWeight: '#f04' }),
    ).toMatchSnapshot();
  });
});
