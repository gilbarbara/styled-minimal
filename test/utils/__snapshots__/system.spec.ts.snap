// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorStyles should return the proper color 1`] = `
Array [
  "
    color: ",
  "#fff",
  ";
  ",
]
`;

exports[`colorStyles should return the proper color 2`] = `
Array [
  "
    color: ",
  "#000",
  ";
  ",
]
`;

exports[`colorStyles should return the proper color 3`] = `
Array [
  "
    color: ",
  "#fff",
  ";
  ",
]
`;

exports[`sx should handle the css prop 1`] = `
Object {
  "&:hover": Object {
    "color": "#333",
  },
  "@media screen and (min-width: 360px)": Object {
    "color": "#999",
  },
  "backgroundColor": "#000",
  "color": "#f04",
}
`;

exports[`sx should handle the responsive prop 1`] = `
Object {
  "@media screen and (min-width: 768px)": Object {
    "color": "#999",
  },
  "color": "#ccc",
}
`;

exports[`system should handle props 1`] = `
Object {
  "@media screen and (min-width: 360px)": Object {
    "fontSize": 22,
  },
  "@media screen and (min-width: 400px)": Object {
    "color": "#000",
  },
  "color": "#333",
  "fontSize": 20,
  "fontWeight": "#f04",
}
`;

exports[`variantStyles should return the proper styles 1`] = `
Array [
  "
    background-color: ",
  "#4B515D",
  ";
    border: ",
  "1px solid #4B515D",
  ";
    color: ",
  "#fff",
  ";
  ",
]
`;

exports[`variantStyles should return the proper styles 2`] = `
Array [
  "
    background-color: ",
  "#FFEB3B",
  ";
    border: ",
  "1px solid #FFEB3B",
  ";
    color: ",
  "#000",
  ";
  ",
]
`;

exports[`variantStyles should return the proper styles 3`] = `
Array [
  "
    background-color: ",
  "#2E2E2E",
  ";
    border: ",
  "1px solid #2E2E2E",
  ";
    color: ",
  "#fff",
  ";
  ",
]
`;
